Add filters to customize behavior for routes

nabsul opened the pull request:
https://github.com/mac-/hapi-statsd/issues/17

This change adds filtering to the hapi-statsd plugin. For our purposes it will allow us to disable stats by default and then customize the stat names on whatever endpoint we decide to use.

Once this is reviewed and merged to master, we can test it out with our server. After that I'll request to get this merged into the original repo.

These changes should also maintain backwards compatibility with users of the plugin who are not aware or need these features.

To test:
- Review the code and unit tests
- You could in theory install this with `npm install [local path to this code]` into the `add/statsd` server branch.

@allendav @jeffstieler @DanReyLop @robobot3000

nabsul committed:
Add filters to customize behavior for routes

nabsul committed:
Add comment to clarify the role of the filterCache

nabsul committed:
Fix up some of the spacing for more consistent coding style

DanReyLop commented:
From the `README`:

> Don't forget to bump the version in the package.json using the semver spec as a guide for which part to bump. Submit a pull request when your work is complete.

Since this is a new, backwards-compatible feature, this should be a `minor` version, so `5.1.0`.

DanReyLop commented on lib/hapi-statsd.js:
```diff
@@ -19,13 +24,36 @@ module.exports.register = function (server, options, next) {
+				return (filter.path || filter.method || filter.status) &&
+					(!filter.path || route.path === filter.path) &&
+					(!filter.method || route.method.toUpperCase() === filter.method.toUpperCase()) &&
+					(!filter.status || status === filter.status);
```
This logic is quite hard to read. I would gladly trade a few extra lines of code for a well-structured `if / else if / else` block.

DanReyLop commented on lib/hapi-statsd.js:
```diff
@@ -19,13 +24,36 @@ module.exports.register = function (server, options, next) {
 	server.ext('onPreResponse', function (request, reply) {
 		var startDate = new Date(request.info.received);
 		var statusCode = (request.response.isBoom) ? request.response.output.statusCode : request.response.statusCode;
+		var method = request.method.toUpperCase();
```
This `var` is pointless here, you're only using its value in one place.

nabsul replied:
I'm keeping with the coding standard of the original module. They don't use const or let.

DanReyLop replied:
I mean that there's no advantage of defining the variable here. You can just use `request.method.toUpperCase()` directly (as it was before).

nabsul commented:
I just realized that I didn't have code coverage working correctly. I'm going to need to continue working on this.

nabsul committed:
Add another test to improve code coverage

nabsul commented:
Alrighty. I've fixed up the code coverage of this change.

DanReyLop commented on test/integration/hapi-statsd.tests.js:
```diff
@@ -33,36 +36,94 @@ beforeEach(function(done) {
+	});
 
-		server.inject('/test/123', function (res) {
+	it('should use cached value', function(done) {
```
You're not really checking that the cache is being used... this just seems a test for the sake of coverage.

nabsul replied:
That's a good point, but I don't know how I could possibly test this.

nabsul replied:
I guess one question to ask: Is this caching mechanism necessary? Or are we ok iterating through that array of filters on every request?

DanReyLop replied:
I don't know how to test it without introducing mocks or spies.

We shouldn't hold back a useful feature just because we can't figure out how to test it. Caching may be a micro-optimisation for us but for other users of the plugin can be important.

On the other hand, we shouldn't add an useless test just for the sake of coverage. The maintainer will probably have the last word on this, but I would just remove the test and accept that the coverage will decrease 2 lines. Was the coverage 100% before?

nabsul committed:
Add documentation for the new options

nabsul committed:
Increment the version of the package

nabsul committed:
Refactor the filter matching code for better readability
Add unit tests to cover the split up if statements
Rename some vars to avoid naming conflicts/overrides

nabsul committed:
Reword and to or for better accuracy

DanReyLop commented on README.md:
```diff
@@ -81,6 +81,25 @@ The prefix to add to every stat collected. Usually used for grouping a set of st
+* enableTimer: Enable/disable the timer stat from being reported
+* enableCounter: Enable/disable the count stat from being reported
+
+Matches by the route's unique ID. This match takes precedence over all other types of matching
```
Was something supposed to be after these words? I think this line doesn't belong here.

Also, an example would be nice.

nabsul replied:
Great idea. I added an example.

DanReyLop reviewed (approved):
This looks great. I say this PR is more than ready to be submitted. Let's hear what the maintainer has to say :)

nabsul committed:
Add an example of the filtering config
